samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1

```diff
@@ -213,32 +216,50 @@ function Get-AzureRmCachedAccessToken() {
     $ErrorActionPreference = 'Stop'
     Set-StrictMode -Off
 
-    $azureRmProfile = [Microsoft.Azure.Commands.Common.Authentication.Abstractions.AzureRmProfileProvider]::Instance.Profile
-    if (-not $azureRmProfile.Accounts.Count) {
-        Write-Error "Ensure you have logged in before calling this function."
+    if ("AzureAutomation/" -eq $env:AZUREPS_HOST_ENVIRONMENT) {

```

Comment:
When running in Azure Automation environment, we need to get the token another way (doing a direct Rest API call based on a localhost identity endpoint).  This is based on Automation accounts that use Run As, Service Principal, User-provided MSI or System-provided MSI.

Reply:
@petehauge - Code itself for calling the identity endpoint looks good.

However, before we add this, did you test call `Connect-AzAccount -Identity`?  It's the supported cmdlet to connect using an MSI.  See https://docs.microsoft.com/en-us/powershell/module/az.accounts/Connect-AzAccount?view=azps-7.3.2#example-5--connect-using-a-managed-service-identity.   

If running, `Connect-AzAccount -Identity` doesn't work, then code looks good to me.

Reply:
Yes-  I agree that's the way to login - the user would need to call Connect-AzAccount -Identity in their runbook (I tested that this works), but in the library we specifically need the access token which isn't available after calling the commandlet to login with an MSI...  I'll try again using Connect-AzAccount and see if the access token is populated and let you know!
https://docs.microsoft.com/en-us/azure/automation/enable-managed-identity-for-automation

Reply:
Sounds good.  I just remember having a similar issue with the Azure PowerShell functions and that call fixed the issue.  

If that does work, I wonder what the best way to educate people would be.

Reply:
@EMaher  - good call on this, I tried with simplified code relying on Connect-AzAccount -Identity and it all worked as expected!  I've simplified the code and will get it checked in shortly after some additional testing.